File: frontend/src/app/services/apartment/apartment.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpErrorResponse, HttpParams } from '@angular/common/http';
import { Observable, throwError } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { ApartmentDTO } from '../../dtos/apartment.dto';
import { ApartmentRequestDTO } from '../../dtos/apartmentRequest.dto';

const BASE_URL = '/api/v1/apartments';

@Injectable({ providedIn: 'root' })
export class ApartmentService {
  constructor(private httpClient: HttpClient) {}

  public getAllApartments(page: number = 0, pageSize: number = 10): Observable<ApartmentDTO[]> {
    const params = new HttpParams().set('page', page).set('pageSize', pageSize);
    return this.httpClient
      .get<ApartmentDTO[]>(`${BASE_URL}/`, { params })
      .pipe(catchError((error) => this.handleError(error)));
  }

  public getApartmentById(id: number): Observable<ApartmentDTO> {
    return this.httpClient
      .get<ApartmentDTO>(`${BASE_URL}/${id}`)
      .pipe(catchError((error) => this.handleError(error)));
  }

  public createApartment(apartment: ApartmentRequestDTO): Observable<ApartmentDTO> {
    const formData = this.buildFormData(apartment);
    return this.httpClient
      .post<ApartmentDTO>(BASE_URL, formData)
      .pipe(catchError((error) => this.handleError(error)));
  }

  public updateApartment(id: number, apartment: ApartmentRequestDTO): Observable<ApartmentDTO> {
    const formData = this.buildFormData(apartment);
    return this.httpClient
      .put<ApartmentDTO>(`${BASE_URL}/${id}`, formData)
      .pipe(catchError((error) => this.handleError(error)));
  }

  public deleteApartment(id: number): Observable<void> {
    return this.httpClient
      .delete<void>(`${BASE_URL}/${id}`)
      .pipe(catchError((error) => this.handleError(error)));
  }

  public searchApartments(options: {
    services?: string[];
    minCapacity?: number;
    startDate?: string;
    endDate?: string;
    page?: number;
    pageSize?: number;
  }): Observable<ApartmentDTO[]> {
    let params = new HttpParams();

    if (options.services) {
      options.services.forEach((s) => (params = params.append('services', s)));
    }
    if (options.minCapacity) params = params.set('minCapacity', options.minCapacity);
    if (options.startDate) params = params.set('startDate', options.startDate);
    if (options.endDate) params = params.set('endDate', options.endDate);

    params = params
      .set('page', options.page ?? 0)
      .set('pageSize', options.pageSize ?? 10);

    return this.httpClient
      .get<ApartmentDTO[]>(`${BASE_URL}/search`, { params })
      .pipe(catchError((error) => this.handleError(error)));
  }

  public getAllServices(): Observable<string[]> {
    return this.httpClient
      .get<string[]>(`${BASE_URL}/services`)
      .pipe(catchError((error) => this.handleError(error)));
  }

  public checkAvailability(
    id: number,
    startDate: string,
    endDate: string
  ): Observable<boolean> {
    const params = new HttpParams().set('startDate', startDate).set('endDate', endDate);
    return this.httpClient
      .get<boolean>(`${BASE_URL}/${id}/availability`, { params })
      .pipe(catchError((error) => this.handleError(error)));
  }

  private buildFormData(dto: ApartmentRequestDTO): FormData {
    const formData = new FormData();
    formData.append('name', dto.name);
    formData.append('description', dto.description);
    formData.append('price', dto.price.toString());
    formData.append('capacity', dto.capacity.toString());

    if (dto.services) {
      dto.services.forEach((service) => formData.append('services', String(service)));
    }
    
    if (dto.images && dto.images.length > 0) {
      dto.images.forEach((image) => formData.append('images', image));
    }
    
    return formData;
  }

  private handleError(error: HttpErrorResponse) {
    console.error('ERROR en ApartmentService:', error);
    return throwError(() => error);
  }
}

File: backend/apartment/src/main/java/com/skyapartments/apartment/service/ApartmentService.java
package com.skyapartments.apartment.service;

import java.time.LocalDate;
import java.util.Collections;
import java.util.Set;

import org.springframework.stereotype.Service;

import com.skyapartments.apartment.dto.ApartmentDTO;
import com.skyapartments.apartment.dto.ApartmentRequestDTO;
import com.skyapartments.apartment.exception.BusinessValidationException;
import com.skyapartments.apartment.exception.ResourceNotFoundException;
import com.skyapartments.apartment.model.Apartment;
import com.skyapartments.apartment.repository.ApartmentRepository;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import jakarta.transaction.Transactional;

@Service
public class ApartmentService {
    
    private final ApartmentRepository apartmentRepository;
    private final ImageService imageService;

    public ApartmentService (ApartmentRepository apartmentRepository, ImageService imageService) {
        this.apartmentRepository = apartmentRepository;
        this.imageService = imageService;
    }

    public Page<ApartmentDTO> getAllApartments(Pageable pageable) {
        return apartmentRepository.findAll(pageable).map(apto -> new ApartmentDTO(apto));
    }

    public ApartmentDTO getApartmentById (Long id) {
        Apartment apartment = apartmentRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Apartment not found"));
        return new ApartmentDTO(apartment);
    }

    @Transactional
    public ApartmentDTO createApartment (ApartmentRequestDTO apartmentRequestDTO) {
        if (apartmentRepository.existsByName(apartmentRequestDTO.getName())) {
            throw new BusinessValidationException("An apartment with this name already exists");
        }

        Apartment apartment = new Apartment(apartmentRequestDTO);
        apartment = apartmentRepository.save(apartment);

        // Upload image
        if (apartmentRequestDTO.getImage() != null && !apartmentRequestDTO.getImage().isEmpty()) {
            try {
                String imageUrl = imageService.saveImage(apartmentRequestDTO.getImage(), apartment.getId());
                apartment.setImageUrl(imageUrl);
            } catch (Exception e) {
                throw new RuntimeException("Error uploading image: " + e.getMessage(), e);
            }
            
        }
        return new ApartmentDTO(apartmentRepository.save(apartment));
    }

    @Transactional
    public ApartmentDTO updateApartment(Long id, ApartmentRequestDTO apartmentRequestDTO) {

        Apartment apartment = apartmentRepository.findById(id)
            .orElseThrow(() -> new BusinessValidationException("Apartment not found with id " + id));

        apartment.setName(apartmentRequestDTO.getName());
        apartment.setDescription(apartmentRequestDTO.getDescription());
        apartment.setPrice(apartmentRequestDTO.getPrice());
        apartment.setCapacity(apartmentRequestDTO.getCapacity());
        apartment.setServices(apartmentRequestDTO.getServices());

        // Delete old image
        imageService.deleteImage(apartment.getImageUrl());
    
        apartment.setImageUrl(null);

        // Upload new images
        if (apartmentRequestDTO.getImage() != null) {
            try {
                String imageUrl = imageService.saveImage(apartmentRequestDTO.getImage(), apartment.getId());
                apartment.setImageUrl(imageUrl);
            } catch (Exception e) {
                    throw new RuntimeException("Error uploading image: " + e.getMessage(), e);
            }
        }
        return new ApartmentDTO(apartmentRepository.save(apartment));
    }

    public void deleteApartment(Long id) {
        Apartment apartment = apartmentRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Apartment not found"));

        // Delete images from S3
        if (apartment.getImageUrl() != null) {
            imageService.deleteImage(apartment.getImageUrl());
        }
        apartmentRepository.deleteById(id);
    }
    
    public Page<ApartmentDTO> searchApartments(
            Set<String> services,
            int minCapacity,
            LocalDate startDate,
            LocalDate endDate,
            Pageable pageable
    ) {
        Set<Long> unavailable = Collections.emptySet();
        if (startDate != null && endDate != null && endDate.isBefore(startDate)) {
        throw new BusinessValidationException("End date must be after start date");
        }
        if (startDate != null && endDate != null) {
            //TODO. Conect to Booking MS to get the apartments that are unavailable in that dates
        }

        int serviceCount = (services != null) ? services.size() : 0;

        return apartmentRepository.findAvailableWithOptionalFilters(
                    services,
                    serviceCount,
                    minCapacity,
                    unavailable.isEmpty() ? null : unavailable,
                    pageable
                )
                .map(ApartmentDTO::new);
    }

    public Set<String> getAllServices() {
        return apartmentRepository.findDistinctServices();
    }

    public Boolean checkAvailability (Long id, LocalDate startDate, LocalDate endDate) {
        Apartment apartment = apartmentRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Apartment not found"));

        Set<Long> unavailable = Collections.emptySet();
        //TODO. Conect to Booking MS to get the apartments that are unavailable in that dates
        return !unavailable.contains(id);
    }

}
